content/blog/2019/08/2019-08-17-role-strategy-performance-folder-auth.adoc

```diff
@@ -0,0 +1,100 @@
+---
+layout: post
+title: "Performance Improvements to Role Strategy Plugin and Folder Auth Plugin"
+tags:
+- plugins
+- performance
+- gsoc
+- gsoc2019
+author: abhyudayasharma
+---
+
+The task for my link:/projects/gsoc/2019/role-strategy-performance[Google Summer of Code Project]
+was to improve the performance of the plugin:role-strategy[Role Strategy Plugin].
+I had started by creating a link:/blog/2019/06/21/performance-testing-jenkins[framework]
+for running micro-benchmarks for Jenkins plugins. I had started by creating Docker
+image for the simulation of the performance slowdown in the plugin followed by
+profiling by using link:https://visualvm.github.io/[VisualVM]. Once we had
+determined the hot-spots, I started creating benchmarks for common and some
+user-provided configurations of the plugin.
+
+= Role Strategy Plugin - improving the performance
+
+A couple of major changes were made to the Role Strategy Plugin to improve its
+performance. First, we started collection of roles that matched a given
+project name. The Role Strategy plugin before version 2.12 used to run over
+regular expressions for every role that it had for every permission checking
+request it got. Storing this produced set of roles in the memory provides us
+large improvements in performance and avoids repeated matching of project names
+with regular expressions. For keeping the plugin working securly, we invalidate
+the cache whenever any update is made to the roles.
+
+Jenkins' permission model allows one permissions to imply other permissions.
+When a permission check is made, we need to check if the user has any of
+permissions that would imply this permisison. For every permission checking
+request that that the Role Strategy, it used to calculate all the implying
+permissions. To avoid this, we now calculate and store implying permissions
+for every permission in the Jenkins system when the plugin gets loaded.
+
+Permissions like the `RUN_SCRIPTS` permission are considered dangerous because
```

Comment:
How this paragraph is related to the performance improvements? Fine for a side note in the presentation, purpose of it in this context is unclear